eclipse-theia/theia
# Fix URI Aware Command Handler Argument Duplication

Signed-off-by: Colin Grant <colin.grant@ericsson.com>



#### What it does

Fixes #8585 by preventing duplication of arguments passed in. Also addresses the `todo` comment by creating separate classes for mono- and multi-select cases.

There are some remaining questions:
- Given the potential confusion created by the generic and options, it may be desirable to deprecate the old `UriAwareCommandHandler` - except that it's likely to have been used quite widely.
- Creating separate classes for the mono- and multi-select cases seems a bit heavy. Instead it would be possible to encourage users to instantiate the `UriAwareCommandHandler` class through the methods `UriAwareCommandHandler.MultiSelect` and `UriAwareCommandHandler.MonoSelect` (or better names :)), since then the typing only needs to be correct there.
- This is potentially breaking two cases:
    1. If anyone was working around it and expecting two copies of the URI.
    1. The old implementation would turn a single URI into `[URI]` in a multi-select context, but that seems like bad behavior. If the function expects URI[], then it seems like URI is a bad call, not something to be corrected. But if anyone was relying on that, they would now get an unexpected result.

#### How to test


1. Run the tests.
1. Copy the snippet from #8585 into `packages/core/src/browser/common-frontend-contribution.ts` in the `registerCommands` function, build, and run the 'Will Cause an Error' command. See that it doesn't anymore.

![image](https://user-images.githubusercontent.com/62660806/95127465-a3e3e680-071d-11eb-9caa-31f904fd153f.png)

1. Use some `UriAwareCommandHandler`s and see that they work (e.g. right click on a folder in the file explorer and try 'Find in Folder')

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
